examples/genserver.exs

```diff
@@ -0,0 +1,125 @@
+Mix.install([:mint_web_socket, :castore])
+
+defmodule Ws do
+  use GenServer
+
+  require Logger
+  require Mint.HTTP
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  def send_message(pid, text) do
+    GenServer.call(pid, {:send_text, text})
+  end
+
+  @impl GenServer
+  def handle_call({:send_text, text}, _from, state) do
+    new_state = send_frame({:text, text}, state)
+    {:reply, :ok, new_state}
+  end
+
+  @impl GenServer
+  def init(_opts) do
+    {:ok, conn} = Mint.HTTP.connect(:http, "127.0.0.1", 1234)
+
+    {:ok, conn, ref} =
+      Mint.WebSocket.upgrade(
+        :ws,
+        conn,
+        "/",
+        []
+      )
+
+    # Receive response to the upgrade call above
+    receive do
+      message when Mint.HTTP.is_connection_message(conn, message) ->
+        {:ok, conn, responses} = Mint.WebSocket.stream(conn, message)
+        response = reduce_response(responses, ref)
+        {:ok, conn, websocket} = Mint.WebSocket.new(conn, ref, response.status, response.headers)
+
+        {:ok,
+         %{
+           ref: ref,
+           websocket: websocket,
+           conn: conn
+         }}
+    end
+  end
+
+  @impl GenServer
+  def handle_info(msg, state) do
+    with {:ok, conn, responses} <- Mint.WebSocket.stream(state.conn, msg),
+         state <- %{state | conn: conn},
+         response <- reduce_response(responses, state.ref),
+         {state, message} <- decode_message(response, state),
+         state <- handle_message(message, state) do
+      {:noreply, state}
+    end
+  end
+
+  defp reduce_response(responses, ref) do
+    Enum.reduce(responses, %{}, fn
+      {:status, ^ref, code}, acc ->
+        Map.put(acc, :status, code)
+
+      {:headers, ^ref, headers}, acc ->
+        Map.update(acc, :headers, headers, &(&1 ++ headers))
+
+      {:data, ^ref, data}, acc ->
+        Map.update(acc, :data, data, &(&1 <> data))
+
+      {:done, ^ref}, acc ->
+        Map.put(acc, :done, true)
+    end)
+  end
+
+  defp decode_message(%{status: status, headers: headers}, state) do
+    case Mint.WebSocket.new(state.conn, state.ref, status, headers) do
+      {:ok, conn, websocket} ->
+        {%{state | conn: conn, websocket: websocket}, :connect}
+
+      {:error, _conn, reason} ->
+        raise "Unknown response #{inspect(reason)}"
+    end
+  end
+
+  defp decode_message(message, state) do
+    case Mint.WebSocket.decode(state.websocket, message.data) do
+      {:ok, websocket, frames} ->
+        {%{state | websocket: websocket}, frames}
+
+      {:error, _websocket, reason} ->
+        raise "connection closed #{inspect(reason)}"
```

Comment:
```suggestion
        raise "Failed to decode WebSocket frame: #{inspect(reason)}"
```

`Mint.WebSocket.decode/2` will fail if it can't decode a frame (for example if the server sends a non-spec-conformant frame). If the connection is closed gracefully by the remote, it returns a `{close, close_code, close_reason}` frame tuple.